### PR TITLE
New sample: Display device location with autopan modes

### PR DESCRIPTION
**Pull request by sclaridge**
Here is the Display device location sample! This sample uses the `SimulatedLocationDataSource` to show off the Autopan modes for 100.10.0. There will be a separate sample coming in the future that shows off the `NMEALocationDataSource`. 

Some areas I would like some feedback on:
- Describing the Autopan modes in the How To section of the README inspired by the Android sample README. Alternatively, the descriptions could go in the Additional information section.

- The velocity of the `SimulatedLocationDataSource`. Is it too fast/slow?

-  I have set the iteration rate on the location display in `COMPASS_NAVIGATION` mode. Based on feedback from the Samples team, the desired effect would simulate a user walking and pivoting around each location point. By setting the iteration rate, it sort of simulates this? I can definitely remove it if it just adds extra lines of code. 

- The location points are based on a feature I created in Pro at the Redlands HQ. I can create/use new data if it is too disorienting for the user. 

Thanks! 

@JonLavi

### CODE REVIEW

File: map_view/display-device-location/README.md
Comment (JonLavi) [inline map_view/display-device-location/README.md]: I'm thinking it might be nice to show a more zoomed-in map in the screenshot

File: map_view/display-device-location/README.md
Comment (JonLavi) [inline map_view/display-device-location/README.md]: I wonder if it might be helpful to mention this is simulated data in the 'how to use', even if it's just clarifying something along the lines:
```
Click the checkbox to show the location symbol at a simulated location on the map.
```

Otherwise it's a bit puzzling where that location comes from, both when using the sample or reading the 'how it works' section

File: map_view/display-device-location/README.metadata.json
Comment (JonLavi) [inline map_view/display-device-location/README.metadata.json]: @Rachael-E might be able to tell whether enums/method calls are safe to put in this section, as it is used for generating some linkes on the user webpages?

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationLauncher.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationLauncher.java]: Alan suggested adding a copyright statement to these files in a previous review. Something to consider @Rachael-E @TADraeseke ?

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I'm a bit puzzled why in the app, when setting this to compass, the movement seems to switch to a more step-by-step style rather than a fluid movement of the position. Is this on purpose?

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: @jenmerritt has recently made some of these comments more 'plain english', and I would agree that it makes it a bit nicer:
https://github.com/Esri/arcgis-runtime-samples-java/pull/562/files#diff-233d54d978f8d006706bc3736efd6f05f86d772aff9079b897c9f9fa12c33a88R55

Might be good to update these here too?

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I think it'd be nice to start the sample with the box getting ticked and the location showing, and one of the auto-pan modes enabled (as soon as the map is loaded, of course). That way we're going straight into the action, rather than showing no location at all on startup? Personal preference though, so just proceed as designed.

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I don't think it's necessary to add the listener again every time the checkbox is ticked.
We could simply put this block of code on line 147, and then this listener is added to the combobox on startup. When we enable/disable the combobox through the checkbox, we effectively block the user from changing the autopan mode, and there's no extra listener added every time.

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I see now where the step-by-step movement is coming from. From my perspective this is a little misleading to a user, because they might think using this auto-pan-mode has some sort of effect on how the movement is displayed. Might be worth discussing in a review session, if this design isn't finalised yet?

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: If you move adding the listener to the combobox outside of the interaction with the checkbox, you could do a little bit of refactoring:
```
      // toggle the location display visibility on check
      checkbox.setOnAction(event -> {
        if (checkbox.isSelected()) {
          // start the location display
          locationDisplay.startAsync();

        } else {
          // turn off the location display
          locationDisplay.stop();
        }

        // toggle the combo box interactions
        comboBox.setDisable(!checkbox.isSelected());
      });
```

Since you always want to toggle the `disabled` property pf the combobox in line with the `is selected` property of the checkbox, you can do this outside of the if/else statement:
`comboBox.setDisable(!checkbox.isSelected());`
Might not be everyone's cup of tea though, so I'll leave that to you and @Rachael-E to decide.

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I'd personally have the tickbox ticked here, and put one of the auto-pan-modes on (e.g. recenter), so that the user is taken straight to the action.

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [inline map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: Personally I find it a bit more intuitive to use a streets basemap for navigation, rather than satellite imagery (because it shows roads and road names more clearly), but that's a matter of preference!

**Review by JonLavi — APPROVED** [review]
A very polished sample @sclaridge !
The only major suggestion I have is moving the block
```
 comboBox.getSelectionModel().selectedItemProperty().addListener(e -> {
```
outside of the interaction with the checkbox, and just have this happen when the app starts up. That way there's not a new listener added every time the box is ticked.

Other than that, there's not much added to this sample on top of what 'show location history' does... Maybe it's worth considering merging the two? It'll just be adding the AutoPanMode combobox!


As for your questions:


>    Describing the Autopan modes in the How To section of the README inspired by the Android sample README. Alternatively, the descriptions could go in the Additional information section.

I'd say this is a good place to put it, as it's a key part of `LocationDisplay`.

>    The velocity of the SimulatedLocationDataSource. Is it too fast/slow?

Seems perfect to me!

>    I have set the iteration rate on the location display in COMPASS_NAVIGATION mode. Based on feedback from the Samples team, the desired effect would simulate a user walking and pivoting around each location point. By setting the iteration rate, it sort of simulates this? I can definitely remove it if it just adds extra lines of code.

I am thinking some users might be lead to believe that the different AutoPanMode somehow affects how the location marker moves. Maybe it's better to not change the iteration rate?

>    The location points are based on a feature I created in Pro at the Redlands HQ. I can create/use new data if it is too disorienting for the user.

Seems perfect, it's almost tradtion to use the esri campus as a sample dataset 👍 

I'll approve so that you can move on with getting a second review once you have made the changes that suit you 👍

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (JonLavi) [reply map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I'm noticing another thing here, a better variable name would be `checkBox`.

And another alternative for enabling/disabling the checkbox and combobox in tandem would be to use something called property binding:
```
      checkbox.selectedProperty().bindBidirectional(comboBox.disableProperty());
```

It's quite a useful concept, so worth looking into a little!

File: map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java
Comment (sclaridge) [reply map_view/display-device-location/src/main/java/com/esri/samples/display_device_location/DisplayDeviceLocationSample.java]: I think this is a good point. I will show this off at the samples review.

**Comment by sclaridge** [comment]
Thanks for the feedback @JonLavi! 
> Other than that, there's not much added to this sample on top of what 'show location history' does... Maybe it's worth considering merging the two? It'll just be adding the AutoPanMode combobox!

It looks like this sample will be renamed "Display device location with autopan modes", and the Show location history sample will show off how to handle the location changed events.

**Comment by sclaridge** [comment]
The sample has been renamed! The sample now starts off in `RECENTER` mode. Based on feedback from the samples team, I have removed the iteration rate code. 

I will also wait to retake the new screenshot until the issue with the symbol size has been resolved.  

@Rachael-E

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: I think a wee gif here would look great: the static image doesn't show a user much about what the sample does other than its UI set up.

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: The sample starts with the checkbox already selected, we can reorder this a little: 

```suggestion
The sample loads with a symbol marking a simulated location on the map. Explore the available autopan modes by selecting an option from the drop down box :
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md:18]:
```suggestion
Uncheck the box to stop the location display. 
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: since we've already mentioned the class map view above, we can refer to it as plain english afterwards: 
```suggestion
2. Get the `LocationDisplay` object by calling `getLocationDisplay()` on the map view.
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]:
```suggestion
3. Start the location display with `startAsync()` to begin receiving location updates.
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: this behaviour is set on the map view, not the map: let's update this text to reflect that
```suggestion
5. Use `locationDisplay.setAutoPanMode()` to change how the map view behaves when location updates are received.
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md:15]: let's rename this in the drop down to match this notation (currently the drop down is "recenter")

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: I'd argue this isn't a relevant API for this sample: ArcGISMap is relevant in ALL 2D samples, but unless it's forming the main part of a sample (like "Display a Map, set zoom scale, set location etc) we don't need to include it.

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: we are in Map view for this sample :) is there a problem with the script?
```suggestion
    "category": "Map view",
```

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: this doesn't match the description in the above readme?

```suggestion
    "description": "Display your position on the map and switch between different types of autopan modes.",
```

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: ok maybe looks like this metadata file hasn't been updated since last review changes? seems to be a few outdated things including the .png. I'll re-review this once it's gone through the script again :) 
```suggestion
        "DisplayDeviceLocationWithAutopanModes.png"
```

File: map_view/display-device-location-with-autopan-modes/README.md
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.md]: the location display is on the map view not the map, I wonder if map view is better as a tag here. But probably best to go back to the design author for this, and it shouldn't block this sample being merged.

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: eek definitely not :D 
```suggestion
    "redirect_from":  "",
```

File: map_view/display-device-location-with-autopan-modes/settings.gradle
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/settings.gradle:1]: the name for this project is the following no?

```suggestion
rootProject.name = 'display-device-location-with-autopan-modes'
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]:
```suggestion
      // create a map view and add the map to it
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]: since this is quite a bunch of UI steps up front that aren't directly related to the important API parts of this sample, I'd be happy to move these into a `setupUI` method.  So the user can get straight to the important bits.

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java:100]: let's try out `var` again here: there are A LOT of simulatedlocationdatasources here and this is exactly the sort of thing local variable type inference was introduced to help with readability. It's obvious from variable name that it is a simulated location data source so is a good `var` candidate.
 
```suggestion
      var simulatedLocationDataSource = new SimulatedLocationDataSource();
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java:106]: we could also opt for `var` here:

```suggestion
      var locationDisplay = mapView.getLocationDisplay();
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]: these comments aren't entirely accurate as to what the API is doing. 

Whilst `startAsync()` and `stop()` do indeed show/hide the location display, what it's doing at an API level is initializing and then receiving location data source updates. When stopping, it ceases to receive the location updates. 

I'd therefore suggest the following: 

```suggestion
      // control location display updates and visibility
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]:
```suggestion
          // start receiving location updates and display the current location with a default round blue symbol
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]:
```suggestion
          // stop receiving location updates and displaying location symbol 
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]: just to keep this naming in keeping with how the API documentation has it and the readme. 
```suggestion
          case "Re-center":
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]: since this is what happens right after loading the sample, it makes sense to move this just after the locationDisplay initializer steps (and before the checkBox on action events).

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java:170]: something I noticed is that when clicking and dragging the map, the autopan mode stops but remains stuck on e.g. compass without being able to activate it again without clicking to another mode, and back to compass. 

We could add in a mapview mouse listener whereby if a user clicks on the map, it will remove the autopan mode by selecting "Off" from the combobox with something like follows: open to other suggestions here too if you think there is a better solution! 

```java
      mapView.setOnMousePressed(event -> {
        if (event.getButton() == MouseButton.PRIMARY) {
          comboBox.setValue("Off");
        }
      });
```

**Review by Rachael-E — CHANGES_REQUESTED** [review]
thanks @sclaridge another neat looking sample! I've made a few suggestions to improve documentation and one user interaction thought: let me know if you have any questions/alternate suggestions to what I've suggested!

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (sclaridge) [reply map_view/display-device-location-with-autopan-modes/README.metadata.json]: Yes, I forgot to run the script again after the rename!

**Comment by sclaridge** [comment]
@Rachael-E I've gone ahead and fixed up the README and the README.metadata.json. I've moved the `map.addDoneLoadingListener()`, fixed up some of the comments and the "Re-Center" option in the drop down box. 

**What's new:** I've added the mapview mouse listener to change the combo box when the user scrolls away from the location display. 

Thanks!

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: map is still showing here: is the script not working?

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/README.metadata.json]: I wonder if the script for generating this is broken again - can you check please? @JonLavi might be able to help. 

We must *not* have an empty string within an array: it will break the website build, so this small change is important. it needs to be: 

```
"redirect_from": "",
```

File: map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/src/main/java/com/esri/samples/display_device_location_with_autopan_modes/DisplayDeviceLocationWithAutopanModesSample.java]: nearly! With mouse pressed, a user can press and then click away, and the drop down box won't be updated. 

This is because when the user presses, the locationDisplay mode is still active. When they release (click), the locationDisplay is off but the comboBox still shows the results of the first press. The second press picks up the locationDisplay off, and then the release (click) updates the combobox). 

We can simply change this to `setOnMouseClicked` so that when the user clicks, it updates the combobox correctly. 

```suggestion
      mapView.setOnMouseClicked(event -> {
```

**Review by Rachael-E — CHANGES_REQUESTED** [review]
thanks @sclaridge, some very small but important changes here: thanks for updating these!

File: map_view/display-device-location-with-autopan-modes/build.gradle
Comment (Rachael-E) [inline map_view/display-device-location-with-autopan-modes/build.gradle:1]: just remembered that we should update this gradle script as you are doing on master: to remove deprecated features.

File: map_view/display-device-location-with-autopan-modes/README.metadata.json
Comment (sclaridge) [reply map_view/display-device-location-with-autopan-modes/README.metadata.json]: Hey @Rachael-E , it looks like the README.metadata.json has been updated: https://github.com/Esri/arcgis-runtime-samples-java/blob/0570f41115e9efaa8231a7d71dfff529cb065ad7/map_view/display-device-location-with-autopan-modes/README.metadata.json

**Comment by Rachael-E** [comment]
looks good @sclaridge, though the metadata still looks outdated from my side: I've pulled down from your branch locally but I can still see the "map" reference in keywords and the erroneous "redirect_from" syntax.

**Comment by sclaridge** [comment]
@Rachael-E The README.metadata.json is updated on my side; I'm not too sure what could be the issue!  

I've just removed an unused import and added the correctly sized gif and image.

**Review by Rachael-E — APPROVED** [review]

**Pull request merged**
